File: flutter_custom_tabs_android/android/src/main/java/com/github/droibit/flutter/plugins/customtabs/core/session/CustomTabsSessionManager.java
package com.github.droibit.flutter.plugins.customtabs.core.session;

import static com.droibit.android.customtabs.launcher.CustomTabsIntentHelper.getCustomTabsPackage;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import androidx.browser.customtabs.CustomTabsSession;

import com.github.droibit.flutter.plugins.customtabs.core.options.CustomTabsSessionOptions;

import java.util.HashMap;
import java.util.Map;

public class CustomTabsSessionManager implements CustomTabsSessionProvider {
    private static final @NonNull String TAG = "CustomTabsAndroid";

    private final @NonNull Map<String, CustomTabsSessionController> cachedSessions;

    public CustomTabsSessionManager() {
        this(new HashMap<>());
    }

    @VisibleForTesting
    CustomTabsSessionManager(@NonNull Map<String, CustomTabsSessionController> cachedSessions) {
        this.cachedSessions = cachedSessions;
    }

    public @NonNull CustomTabsSessionOptions createSessionOptions(@Nullable Map<String, Object> options) {
        return new CustomTabsSessionOptions.Builder()
                .setOptions(options)
                .build();
    }

    public @Nullable CustomTabsSessionController createSessionController(
            @NonNull Context context,
            @NonNull CustomTabsSessionOptions options
    ) {
        final Boolean prefersDefaultBrowser = options.getPrefersDefaultBrowser();
        final String customTabsPackage = getCustomTabsPackage(
                context,
                prefersDefaultBrowser != null && !prefersDefaultBrowser,
                options.getAdditionalCustomTabs(context)
        );
        if (customTabsPackage == null) {
            return null;
        }

        final CustomTabsSessionController cachedController = cachedSessions.get(customTabsPackage);
        if (cachedController != null) {
            return cachedController;
        }

        final CustomTabsSessionController newController = new CustomTabsSessionController(customTabsPackage);
        cachedSessions.put(customTabsPackage, newController);
        return newController;
    }

    public @Nullable CustomTabsSessionController getSessionController(@NonNull String packageName) {
        return cachedSessions.get(packageName);
    }

    @Override
    public @Nullable CustomTabsSession getSession(@Nullable String packageName) {
        if (packageName == null) {
            return null;
        }

        final CustomTabsSessionController controller = cachedSessions.get(packageName);
        if (controller == null) {
            return null;
        }
        return controller.getSession();
    }

    public void invalidateSession(@NonNull String packageName) {
        final CustomTabsSessionController controller = cachedSessions.get(packageName);
        if (controller == null) {
            return;
        }
        controller.unbindCustomTabsService();
        cachedSessions.remove(packageName);
    }

    public void handleActivityChange(@Nullable Context activity) {
        Log.d(TAG, "handleActivityChange: " + activity);
        for (CustomTabsSessionController controller : cachedSessions.values()) {
            if (activity == null) {
                controller.unbindCustomTabsService();
            } else {
                controller.bindCustomTabsService(activity);
            }
        }
    }
}

File: flutter_custom_tabs_android/android/src/main/kotlin/com/github/droibit/flutter/plugins/customtabs/core/session/CustomTabsSessionManager.kt
package com.github.droibit.flutter.plugins.customtabs.core.session

import android.content.Context
import androidx.annotation.VisibleForTesting
import androidx.browser.customtabs.CustomTabsSession
import com.droibit.android.customtabs.launcher.getCustomTabsPackage
import com.github.droibit.flutter.plugins.customtabs.core.options.CustomTabsSessionOptions

class CustomTabsSessionManager @VisibleForTesting internal constructor(
  private val cachedSessions: MutableMap<String, CustomTabsSessionController>
) : CustomTabsSessionProvider {
  constructor() : this(mutableMapOf())

  fun createSessionOptions(options: Map<String, Any>?): CustomTabsSessionOptions {
    return CustomTabsSessionOptions.Builder()
      .setOptions(options)
      .build()
  }

  fun createSessionController(
    context: Context,
    options: CustomTabsSessionOptions
  ): CustomTabsSessionController? {
    val prefersDefaultBrowser = options.prefersDefaultBrowser
    val customTabsPackage = getCustomTabsPackage(
      context,
      ignoreDefault = prefersDefaultBrowser != true,
      additionalCustomTabs = options.getAdditionalCustomTabs(context)
    ) ?: return null

    return cachedSessions[customTabsPackage]
      ?: CustomTabsSessionController(customTabsPackage).also {
        cachedSessions[customTabsPackage] = it
      }
  }

  fun getSessionController(packageName: String): CustomTabsSessionController? {
    return cachedSessions[packageName]
  }

  override fun getSession(packageName: String?): CustomTabsSession? {
    return packageName?.let { cachedSessions[it]?.session }
  }

  fun invalidateSession(packageName: String) {
    val controller = cachedSessions[packageName] ?: return
    controller.unbindCustomTabsService()
    cachedSessions.remove(packageName)
  }

  fun handleActivityChange(activity: Context?) {
    for (controller in cachedSessions.values) {
      if (activity == null) {
        controller.unbindCustomTabsService()
      } else {
        controller.bindCustomTabsService(activity)
      }
    }
  }
}
